pkg/puller/puller.go

```diff
@@ -0,0 +1,76 @@
+package puller
+
+import (
+	"log"
+)
+
+var (
+	p *Puller
+)
+
+type Puller struct {
+	// TODO: Should this be a syncmap?
+	ChannelMap map[string]Channel
```

Comment:
I think we need to use sync.Map here: https://golang.org/pkg/sync/#Map not syncmap. syncmap is just a prototype for sync.Map. And yes, I think we should use sync.Map here. The puller has different goroutines to call AddModel and RemoveModel which can read and write the same map concurrently.

Reply:
See the description of sync.Map 
``` Map is like a Go map[interface{}]interface{} but is safe for concurrent use by multiple goroutines without additional locking or coordination. Loads, stores, and deletes run in amortized constant time.

The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

The zero Map is empty and ready for use. A Map must not be copied after first use.
```

Reply:
Ah I meant sync.Map when I wrote that. I just didn’t know if it was necessary to make it a sync.Map. As we are writing to the channel. Not quite convinced it’s necessary, but I can be easily persuaded :)